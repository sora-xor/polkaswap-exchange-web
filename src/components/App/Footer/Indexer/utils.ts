import { IndexerType } from '@soramitsu/soraneo-wallet-web/lib/consts';

export function getIndexerName(type: IndexerType): string {
  return type.charAt(0).toUpperCase() + type.slice(1).toLowerCase();
}
