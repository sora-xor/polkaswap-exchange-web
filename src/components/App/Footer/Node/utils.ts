import { countryCodeEmoji } from 'country-code-emoji';

import store from '@/store';

export function formatLocation(code: string): Nullable<{ flag: string; name?: string }> {
  try {
    const isoCode = code.toUpperCase();
    const flag = countryCodeEmoji(isoCode);
    const location = { flag, name: '' };
    const displayRegions = store.state.settings.displayRegions;
    if (!displayRegions) {
      return location;
    }
    const name = displayRegions.of(isoCode);
    if (name) {
      location.name = name;
    }
    return location;
  } catch (error) {
    console.warn('Unsupported format of location in env.json', error);
    return null;
  }
}
