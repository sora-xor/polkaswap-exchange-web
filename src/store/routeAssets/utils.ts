import { FPNumber, Operation } from '@sora-substrate/util/build';
import { Asset, AccountAsset } from '@sora-substrate/util/build/assets/types';
import { api } from '@soramitsu/soraneo-wallet-web';
import { FiatPriceObject } from '@soramitsu/soraneo-wallet-web/lib/services/subquery/types';

import { Recipient } from './types';

import type { WhitelistArrayItem } from '@sora-substrate/util/build/assets/types';

export default {
  validate(recipient: Recipient) {
    return (
      this.wallet(recipient.wallet) &&
      this.asset(recipient.asset) &&
      this.usd(recipient.usd) &&
      this.amount(recipient.amount)
    );
  },

  name(name: Nullable<string>) {
    return name;
  },

  wallet(wallet: string) {
    return api.validateAddress(wallet);
  },

  asset(asset: Asset | AccountAsset) {
    return asset;
  },

  usd(usd: Nullable<number>) {
    return usd && !isNaN(usd);
  },

  amount(amount: Nullable<number>) {
    return amount && !isNaN(amount) && Number.isFinite(amount);
  },
};

export function getTokenEquivalent(
  priceObject: FiatPriceObject,
  asset: Asset | AccountAsset | WhitelistArrayItem,
  usd: number | string
): FPNumber {
  return new FPNumber(usd).div(FPNumber.fromCodecValue(priceObject[asset.address], asset.decimals));
}

export function getAssetUSDPrice(asset, fiatPriceObject) {
  if (!asset) return null;
  return FPNumber.fromCodecValue(fiatPriceObject[asset.address], asset.decimals);
}
