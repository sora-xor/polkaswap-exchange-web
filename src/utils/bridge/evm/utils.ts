import { Operation } from '@sora-substrate/sdk';

import { evmBridgeApi } from '@/utils/bridge/evm/api';

import type { EvmHistory } from '@sora-substrate/sdk/build/bridgeProxy/evm/types';

export const isOutgoingTx = (tx: EvmHistory): boolean => {
  return tx.type === Operation.EvmOutgoing;
};

export const isUnsignedTx = (tx: EvmHistory): boolean => {
  if (isOutgoingTx(tx)) {
    return !tx.blockId && !tx.txId;
  } else {
    return !tx.externalHash;
  }
};

export const getTransaction = (id: string): EvmHistory => {
  const tx = evmBridgeApi.getHistory(id) as EvmHistory;

  if (!tx) throw new Error(`[Bridge]: Transaction is not exists: ${id}`);

  return tx;
};

export const updateTransaction = (id: string, params = {}): void => {
  const tx = getTransaction(id);
  evmBridgeApi.saveHistory({ ...tx, ...params });
};
