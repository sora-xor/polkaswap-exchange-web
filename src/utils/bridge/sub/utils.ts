import { FPNumber, Operation } from '@sora-substrate/util';

import { subBridgeApi } from '@/utils/bridge/sub/api';
import { SubTransferType } from '@/utils/bridge/sub/types';

import type { CodecString, WithConnectionApi } from '@sora-substrate/util';
import type { RegisteredAccountAsset } from '@sora-substrate/util/build/assets/types';
import type { SubNetwork, SubHistory } from '@sora-substrate/util/build/bridgeProxy/sub/types';

export const isOutgoingTx = (tx: SubHistory): boolean => {
  return tx.type === Operation.SubstrateOutgoing;
};

export const isUnsignedTx = (tx: SubHistory): boolean => {
  const signId =
    subBridgeApi.isEvmAccount(tx.externalNetwork as SubNetwork) && !isOutgoingTx(tx) ? tx.externalHash : tx.txId;

  return !tx.blockId && !signId;
};

export const getTransaction = (id: string): SubHistory => {
  const tx = subBridgeApi.getHistory(id) as SubHistory;

  if (!tx) throw new Error(`[Bridge]: Transaction is not exists: ${id}`);

  return tx;
};

export const updateTransaction = (id: string, params = {}): void => {
  const tx = getTransaction(id);
  const data = { ...tx, ...params };
  subBridgeApi.saveHistory(data);
};

export const determineTransferType = (network: SubNetwork) => {
  if (subBridgeApi.isSoraParachain(network)) {
    return SubTransferType.SoraParachain;
  } else if (subBridgeApi.isRelayChain(network)) {
    return SubTransferType.Relaychain;
  } else if (subBridgeApi.isStandalone(network)) {
    return SubTransferType.Standalone;
  } else {
    return SubTransferType.Parachain;
  }
};

export const isEvent = (e, section: string, method: string) => {
  return e.event.section === section && e.event.method === method;
};

export const isQueueMessage = (e) =>
  isEvent(e, 'messageQueue', 'Processed') ||
  isEvent(e, 'xcmpQueue', 'Success') ||
  isEvent(e, 'xcmpQueue', 'Fail') ||
  isEvent(e, 'dmpQueue', 'ExecutedDownward');

export const isParaInclusion = (e) => isEvent(e, 'paraInclusion', 'CandidateIncluded');

export const isXcmPalletAttempted = (e) => isEvent(e, 'xcmPallet', 'Attempted');

export const isTransactionFeePaid = (e) => isEvent(e, 'transactionPayment', 'TransactionFeePaid');

export const isBridgeProxyUpdate = (e) => isEvent(e, 'bridgeProxy', 'RequestStatusUpdate');

export const getBridgeProxyHash = (events: Array<any>): string => {
  const bridgeProxyEvent = events.find((e) => isBridgeProxyUpdate(e));

  if (!bridgeProxyEvent) {
    throw new Error(`Unable to find "bridgeProxy.RequestStatusUpdate" event`);
  }

  return bridgeProxyEvent.event.data[0].toString();
};

export const isBridgeProxyHash = (e: any, hash: string): boolean => {
  if (!isBridgeProxyUpdate(e)) return false;

  const proxyHash = e.event.data[0].toString();

  return hash === proxyHash;
};

export const getDepositedBalance = (events: Array<any>, to: string, chainApi: WithConnectionApi): [string, number] => {
  const recipient = chainApi.formatAddress(to).toLowerCase();

  const index = events.findIndex((e) => {
    let eventRecipient = '';

    if (isEvent(e, 'balances', 'Deposit') || isEvent(e, 'balances', 'Minted') || isEvent(e, 'tokens', 'Deposited')) {
      eventRecipient = e.event.data.who.toString();
    } else if (isEvent(e, 'assets', 'Transfer')) {
      eventRecipient = e.event.data[1].toString();
    } else if (isEvent(e, 'assets', 'Issued')) {
      eventRecipient = e.event.data.owner.toString();
    }

    if (!eventRecipient) return false;

    const formatted = chainApi.formatAddress(eventRecipient).toLowerCase();

    return formatted === recipient;
  });

  if (index === -1) throw new Error(`Unable to find balance deposit like event`);

  const event = events[index];
  const balance = event.event.data.amount?.toString() ?? event.event.data[3].toString();

  return [balance, index];
};

export const getReceivedAmount = (sendedAmount: string, receivedAmount: CodecString, decimals?: number) => {
  const sended = new FPNumber(sendedAmount, decimals);
  const received = FPNumber.fromCodecValue(receivedAmount, decimals);
  const amount2 = received.toString();
  const transferFee = sended.sub(received).toCodecString();

  return { amount: amount2, transferFee };
};

export const getParachainSystemMessageHash = (events: Array<any>) => {
  const parachainSystemEvent = events.find(
    (e) => isEvent(e, 'parachainSystem', 'UpwardMessageSent') || isEvent(e, 'xcmpQueue', 'XcmpMessageSent')
  );

  if (!parachainSystemEvent) {
    throw new Error(`Unable to find "parachainSystem.UpwardMessageSent" event`);
  }

  return parachainSystemEvent.event.data[0].toString();
};

export const isMessageAccepted = (e) => isEvent(e, 'substrateBridgeOutboundChannel', 'MessageAccepted');

export const getMessageAcceptedNonces = (events: Array<any>): [number, number] => {
  const messageAcceptedEvent = events.find((e) => isMessageAccepted(e));

  if (!messageAcceptedEvent) {
    throw new Error('Unable to find "substrateBridgeOutboundChannel.MessageAccepted" event');
  }

  const batchNonce = messageAcceptedEvent.event.data[1].toNumber();
  const messageNonce = messageAcceptedEvent.event.data[2].toNumber();

  return [batchNonce, messageNonce];
};

export const isMessageDispatched = (e) => isEvent(e, 'substrateDispatch', 'MessageDispatched');

export const getMessageDispatchedNonces = (events: Array<any>): [number, number] => {
  const messageDispatchedEvent = events.find((e) => isMessageDispatched(e));

  if (!messageDispatchedEvent) {
    throw new Error('Unable to find "substrateDispatch.MessageDispatched" event');
  }

  const { batchNonce, messageNonce } = messageDispatchedEvent.event.data[0];
  const eventBatchNonce = batchNonce.unwrap().toNumber();
  const eventMessageNonce = messageNonce.toNumber();

  return [eventBatchNonce, eventMessageNonce];
};

export const isMessageDispatchedNonces = (sendedBatchNonce: number, sendedMessageNonce: number, e: any): boolean => {
  if (!isMessageDispatched(e)) return false;

  const { batchNonce, messageNonce } = e.event.data[0];

  const eventBatchNonce = batchNonce.unwrap().toNumber();
  const eventMessageNonce = messageNonce.toNumber();

  if (eventBatchNonce > sendedBatchNonce) {
    throw new Error(
      `Parachain channel batch nonce ${eventBatchNonce} is larger than tx batch nonce ${sendedBatchNonce}`
    );
  }

  return sendedBatchNonce === eventBatchNonce && sendedMessageNonce === eventMessageNonce;
};

export const isAssetAddedToChannel = (
  e: any,
  asset: RegisteredAccountAsset,
  to: string,
  sended: CodecString,
  chainApi: WithConnectionApi
): boolean => {
  if (!isEvent(e, 'xcmApp', 'AssetAddedToChannel')) return false;

  const { amount, assetId, recipient } = e.event.data[0].asTransfer;
  // address check
  if (chainApi.formatAddress(recipient.toString()) !== chainApi.formatAddress(to)) return false;
  // asset check
  if (assetId.toString() !== asset.address) return false;
  // amount check
  // [WARNING] Implementation could be changed: sora parachain doesn't spent xcm fee from amount
  if (amount.toString() !== sended) return false;

  return true;
};

// Liberland
export const isSoraBridgeAppBurned = (
  e: any,
  asset: RegisteredAccountAsset,
  from: string,
  to: string,
  sended: CodecString,
  chainApi: WithConnectionApi
) => {
  if (!isEvent(e, 'soraBridgeApp', 'Burned')) return false;

  const [networkIdCodec, assetIdCodec, senderCodec, recipientCodec, amountCodec] = e.event.data;

  if (!(networkIdCodec.isMainnet && recipientCodec.isSora)) return false;

  const sender = senderCodec.toString();
  const recipient = recipientCodec.asSora.toString();
  const assetId = assetIdCodec.isLld ? '' : assetIdCodec.asAsset.toString();
  const amount = amountCodec.toString();

  // address check
  if (chainApi.formatAddress(sender) !== chainApi.formatAddress(from)) return false;
  if (chainApi.formatAddress(recipient) !== chainApi.formatAddress(to)) return false;
  // asset check
  if (assetId !== asset.externalAddress) return false;
  // amount check
  if (amount !== sended) return false;

  return true;
};
