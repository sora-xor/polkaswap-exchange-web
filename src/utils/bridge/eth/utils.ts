import { Operation, FPNumber } from '@sora-substrate/sdk';
import { BridgeTxStatus } from '@sora-substrate/sdk/build/bridgeProxy/consts';
import { EthCurrencyType, EthAssetKind } from '@sora-substrate/sdk/build/bridgeProxy/eth/consts';

import { SmartContractType, KnownEthBridgeAsset, SmartContracts } from '@/consts/evm';
import { asZeroValue } from '@/utils';
import { ethBridgeApi } from '@/utils/bridge/eth/api';
import ethersUtil from '@/utils/ethers-util';

import type { RegisteredAccountAsset } from '@sora-substrate/sdk/build/assets/types';
import type { EthHistory, EthApprovedRequest } from '@sora-substrate/sdk/build/bridgeProxy/eth/types';
import type { Subscription } from 'rxjs';

type EthTxParams = {
  asset: RegisteredAccountAsset;
  value: string;
  recipient: string;
  getContractAddress: (symbol: KnownEthBridgeAsset) => Nullable<string>;
  request?: EthApprovedRequest;
};

export const isOutgoingTx = (tx: EthHistory): boolean => {
  return tx.type === Operation.EthBridgeOutgoing;
};

export const isUnsignedFromPart = (tx: EthHistory): boolean => {
  if (isOutgoingTx(tx)) {
    return !tx.blockId && !tx.txId;
  } else {
    return !tx.externalHash;
  }
};

export const isUnsignedToPart = (tx: EthHistory): boolean => {
  if (tx.type === Operation.EthBridgeOutgoing) {
    return !tx.externalHash;
  } else if (tx.type === Operation.EthBridgeIncoming) {
    return false;
  } else {
    return true;
  }
};

export const isUnsignedTx = (tx: EthHistory): boolean => {
  return isUnsignedFromPart(tx);
};

export const getTransaction = (id: string): EthHistory => {
  const tx = ethBridgeApi.getHistory(id) as EthHistory;

  if (!tx) throw new Error(`[Bridge]: Transaction is not exists: ${id}`);

  return tx;
};

export const updateTransaction = async (id: string, params = {}) => {
  const tx = getTransaction(id);
  ethBridgeApi.saveHistory({ ...tx, ...params });
};

export const waitForApprovedRequest = async (tx: EthHistory): Promise<EthApprovedRequest> => {
  const hash = tx.hash;

  if (!hash) throw new Error(`[Bridge]: Tx hash cannot be empty`);
  if (!Number.isFinite(tx.externalNetwork))
    throw new Error(`[Bridge]: Tx externalNetwork should be a number, ${tx.externalNetwork} received`);

  let subscription!: Subscription;

  await new Promise<void>((resolve, reject) => {
    subscription = ethBridgeApi.subscribeOnRequestStatus(hash).subscribe((status) => {
      switch (status) {
        case BridgeTxStatus.Failed:
        case BridgeTxStatus.Frozen:
        case BridgeTxStatus.Broken:
          reject(new Error('[Bridge]: Transaction was failed or canceled'));
          break;
        case BridgeTxStatus.Ready:
          resolve();
          break;
      }
    });
  });

  subscription.unsubscribe();

  const request = await ethBridgeApi.getApprovedRequest(hash);

  if (!request) throw new Error(`[Bridge]: getApprovedRequest is empty, hash="${hash}"`);

  return request;
};

export const waitForIncomingRequest = async (tx: EthHistory): Promise<{ hash: string; blockId: string }> => {
  if (!tx.externalHash) throw new Error('[Bridge]: externalHash cannot be empty!');
  if (!Number.isFinite(tx.externalNetwork))
    throw new Error(`[Bridge]: Tx externalNetwork should be a number, ${tx.externalNetwork} received`);

  let subscription!: Subscription;

  await new Promise<void>((resolve, reject) => {
    subscription = ethBridgeApi.subscribeOnRequest(tx.externalHash as string).subscribe((request) => {
      if (request) {
        switch (request.status) {
          case BridgeTxStatus.Failed:
          case BridgeTxStatus.Frozen:
          case BridgeTxStatus.Broken:
            reject(new Error('[Bridge]: Transaction was failed or canceled'));
            break;
          case BridgeTxStatus.Done:
            resolve();
            break;
        }
      }
    });
  });

  subscription.unsubscribe();

  const soraHash = await ethBridgeApi.getSoraHashByEthereumHash(tx.externalHash as string);
  const soraBlockHash = await ethBridgeApi.getSoraBlockHashByRequestHash(tx.externalHash as string);

  return { hash: soraHash, blockId: soraBlockHash };
};

export async function getIncomingEvmTransactionData({ asset, value, recipient, getContractAddress }: EthTxParams) {
  const isNativeEvmToken = ethersUtil.isNativeEvmTokenAddress(asset.externalAddress);
  const accountId = ethersUtil.accountAddressToHex(recipient);
  const amount = new FPNumber(value, asset.externalDecimals).toCodecString();

  const contractAddress = getContractAddress(KnownEthBridgeAsset.Other) as string;
  const contractAbi = SmartContracts[SmartContractType.EthBridge][KnownEthBridgeAsset.Other];
  const contract = await ethersUtil.getContract(contractAddress, contractAbi);

  const method = isNativeEvmToken ? 'sendEthToSidechain' : 'sendERC20ToSidechain';
  const methodArgs = isNativeEvmToken
    ? [
        accountId, // bytes32 to
      ]
    : [
        accountId, // bytes32 to
        amount, // uint256 amount
        asset.externalAddress, // address tokenAddress
      ];
  const overrides = isNativeEvmToken ? { value: amount } : {};
  const args = [...methodArgs, overrides];

  return {
    contract,
    method,
    args,
  };
}

export async function getOutgoingEvmTransactionData({
  asset,
  value,
  recipient,
  getContractAddress,
  request,
}: EthTxParams) {
  if (!request) throw new Error('request is required!');

  const symbol = asset.symbol as KnownEthBridgeAsset;
  const isValOrXor = [KnownEthBridgeAsset.XOR, KnownEthBridgeAsset.VAL].includes(symbol);
  const bridgeAsset: KnownEthBridgeAsset = isValOrXor ? symbol : KnownEthBridgeAsset.Other;

  const contractAddress = getContractAddress(bridgeAsset) as string;
  const contractAbi = SmartContracts[SmartContractType.EthBridge][bridgeAsset];
  const contract = await ethersUtil.getContract(contractAddress, contractAbi);

  const amount = new FPNumber(value, asset.externalDecimals).toCodecString();

  const isEthereumCurrency = request.currencyType === EthCurrencyType.TokenAddress;
  const bridgeContractMethod = isEthereumCurrency ? 'receiveByEthereumAssetAddress' : 'receiveBySidechainAssetId';

  const method = isValOrXor ? 'mintTokensByPeers' : bridgeContractMethod;

  const args: Array<any> = [
    isValOrXor || isEthereumCurrency
      ? asset.externalAddress // address tokenAddress OR
      : asset.address, // bytes32 assetId
    amount, // uint256 amount
    recipient, // address beneficiary
  ];
  args.push(
    ...(isValOrXor
      ? [
          request.hash, // bytes32 txHash
          request.v, // uint8[] memory v
          request.r, // bytes32[] memory r
          request.s, // bytes32[] memory s
          request.from, // address from
        ]
      : [
          request.from, // address from
          request.hash, // bytes32 txHash
          request.v, // uint8[] memory v
          request.r, // bytes32[] memory r
          request.s, // bytes32[] memory s
        ])
  );

  return {
    contract,
    method,
    args,
  };
}

const gasLimit = {
  approve: BigInt(45000),
  sendERC20ToSidechain: BigInt(53000),
  sendEthToSidechain: BigInt(26093),
  mintTokensByPeers: BigInt(211000),
  receiveByEthereumAssetAddress: {
    ETH: BigInt(155000),
    OTHER: BigInt(181000),
  },
  receiveBySidechainAssetId: BigInt(184000),
};

/**
 * It's in gwei.
 */
const getEthBridgeOutgoingGasLimit = (assetEvmAddress: string, assetKind: EthAssetKind): bigint => {
  switch (assetKind) {
    case EthAssetKind.SidechainOwned:
      return gasLimit.mintTokensByPeers;
    case EthAssetKind.Thischain:
      return gasLimit.receiveBySidechainAssetId;
    case EthAssetKind.Sidechain:
      return ethersUtil.isNativeEvmTokenAddress(assetEvmAddress)
        ? gasLimit.receiveByEthereumAssetAddress.ETH
        : gasLimit.receiveByEthereumAssetAddress.OTHER;
    default:
      throw new Error(`Unknown kind "${assetKind}" for asset "${assetEvmAddress}"`);
  }
};

const getEthBridgeIncomingGasLimit = (assetEvmAddress: string): bigint => {
  return ethersUtil.isNativeEvmTokenAddress(assetEvmAddress)
    ? gasLimit.sendEthToSidechain
    : gasLimit.sendERC20ToSidechain;
};

export async function getEthNetworkFee(
  asset: RegisteredAccountAsset,
  assetKind: string,
  getContractAddress: (symbol: KnownEthBridgeAsset) => Nullable<string>,
  value: string,
  isOutgoing: boolean,
  soraAccount: string,
  evmAccount: string
) {
  let gasLimitTotal!: bigint;

  if (isOutgoing) {
    gasLimitTotal = getEthBridgeOutgoingGasLimit(asset.externalAddress, assetKind as EthAssetKind);
  } else {
    const bridgeContractAddress = getContractAddress(KnownEthBridgeAsset.Other) as string;
    const allowance = await ethersUtil.getAllowance(evmAccount, bridgeContractAddress, asset.externalAddress);
    const approveGasLimit = !!allowance && Number(allowance) < Number(value) ? gasLimit.approve : BigInt(0);

    let txGasLimit!: bigint;

    try {
      if (asZeroValue(value)) {
        throw new Error('Calculation with Zero amount is not allowed');
      }

      const txParams = {
        asset,
        value,
        recipient: soraAccount,
        getContractAddress,
      };
      const { contract, method, args } = await getIncomingEvmTransactionData(txParams);
      const signer = contract.runner;
      const tx = await contract[method].populateTransaction(...args);

      txGasLimit = (await signer?.estimateGas?.(tx)) ?? BigInt(0);
    } catch {
      txGasLimit = getEthBridgeIncomingGasLimit(asset.externalAddress);
    }

    gasLimitTotal = txGasLimit + approveGasLimit;
  }

  const gasPrice = await ethersUtil.getEvmGasPrice();

  return ethersUtil.calcEvmFee(gasPrice, gasLimitTotal);
}
