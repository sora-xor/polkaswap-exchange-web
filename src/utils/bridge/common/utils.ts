import { Operation, isEthOperation, isEvmOperation, isSubstrateOperation } from '@sora-substrate/util';
import { api as soraApi } from '@soramitsu/soraneo-wallet-web';
import { ethers } from 'ethers';

import { isUnsignedTx as isUnsignedEthTx } from '@/utils/bridge/eth/utils';
import { isUnsignedTx as isUnsignedEvmTx } from '@/utils/bridge/evm/utils';
import { isUnsignedTx as isUnsignedSubTx } from '@/utils/bridge/sub/utils';
import ethersUtil from '@/utils/ethers-util';

import type { ApiPromise } from '@polkadot/api';
import type { IBridgeTransaction } from '@sora-substrate/util';
import type { EthHistory } from '@sora-substrate/util/build/bridgeProxy/eth/types';
import type { EvmHistory } from '@sora-substrate/util/build/bridgeProxy/evm/types';
import type { SubHistory } from '@sora-substrate/util/build/bridgeProxy/sub/types';

export const waitForEvmTransactionMined = async (
  tx: ethers.TransactionResponse | null,
  replaceCallback?: (tx: ethers.TransactionResponse | null) => void
): Promise<ethers.TransactionReceipt | null> => {
  if (!tx) throw new Error('[waitForEvmTransactionMined]: tx cannot be empty!');

  try {
    const startBlock = tx.blockNumber ?? (await ethersUtil.getBlockNumber());
    const replaceableTx = tx.replaceableTransaction(startBlock);
    const txReceipt = await replaceableTx.wait();

    return txReceipt;
  } catch (error) {
    if (ethers.isError(error, 'TRANSACTION_REPLACED')) {
      const replacedTx = error.replacement;

      replaceCallback?.(replacedTx);

      return await waitForEvmTransactionMined(replacedTx, replaceCallback);
    }

    throw error;
  }
};

export const getEvmTransactionRecieptByHash = async (
  transactionHash: string
): Promise<{ fee: string; blockHash: string; blockNumber: number; from: string } | null> => {
  try {
    const receipt = await ethersUtil.getEvmTransactionReceipt(transactionHash);

    if (!receipt) throw new Error(`Transaction receipt "${transactionHash}" not found`);

    const { fee, from, blockNumber, blockHash } = receipt;

    return { fee: fee.toString(), blockHash, blockNumber, from };
  } catch (error) {
    return null;
  }
};

export const getBlockEventsByTxIndex = async (blockHash: string, index: number, api: ApiPromise) => {
  const blockEvents = await soraApi.system.getBlockEvents(blockHash, api);
  const transactionEvents = blockEvents.filter(
    ({ phase }) => phase.isApplyExtrinsic && phase.asApplyExtrinsic.toNumber() === index
  );

  return transactionEvents;
};

export const getTransactionEvents = async (blockHash: string, transactionHash: string, api: ApiPromise) => {
  const extrinsics = await soraApi.system.getExtrinsicsFromBlock(blockHash, api);
  const extrinsicIndex = extrinsics.findIndex((ext) => ext.hash.toString() === transactionHash);

  if (extrinsicIndex === -1) throw new Error(`Unable to find extrinsic "${transactionHash}" in block "${blockHash}"`);

  const transactionEvents = await getBlockEventsByTxIndex(blockHash, extrinsicIndex, api);

  return transactionEvents;
};

export const isOutgoingTransaction = (transaction: Nullable<IBridgeTransaction>): boolean => {
  if (!transaction?.type) return false;

  return [Operation.EthBridgeOutgoing, Operation.EvmOutgoing, Operation.SubstrateOutgoing].includes(transaction.type);
};

export const isUnsignedTx = (transaction: Nullable<IBridgeTransaction>): boolean => {
  if (!transaction?.type) return true;

  if (isEthOperation(transaction.type)) return isUnsignedEthTx(transaction as EthHistory);
  if (isEvmOperation(transaction.type)) return isUnsignedEvmTx(transaction as EvmHistory);
  if (isSubstrateOperation(transaction.type)) return isUnsignedSubTx(transaction as SubHistory);

  return true;
};
