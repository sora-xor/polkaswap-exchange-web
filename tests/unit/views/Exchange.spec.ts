import { shallowMount, createLocalVue } from '@vue/test-utils'

import Exchange from '@/views/Exchange.vue'
import { TranslationMock } from '../../utils'

const localVue = createLocalVue()

describe('Exchange.vue', () => {
  beforeEach(() => {
    TranslationMock(Exchange)
  })

  it('should renders correctly', () => {
    const wrapper = shallowMount(Exchange, { localVue })
    expect(wrapper.element).toMatchSnapshot()
  })
})
