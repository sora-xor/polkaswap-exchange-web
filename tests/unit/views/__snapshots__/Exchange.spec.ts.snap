// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Exchange.vue should renders correctly 1`] = `
<div
  class="exchange-container"
>
  <s-tabs
    type="rounded"
  >
    <s-tab
      name="swap"
    >
      <swap-stub />
    </s-tab>
     
    <s-tab
      name="pool"
    >
      <pool-stub />
    </s-tab>
  </s-tabs>
</div>
`;
